### PR TITLE
Connection pool support in connection cache and QUIC connection reliability improvement

### PR DESCRIPTION
**Pull request by lijunwangs**
#### Problem
To improve QUIC transaction send throughput and connection reliability. 

#### Summary of Changes

1. Connection pool is devised which is per address in the connection cache. The count is configurable. The default is 4 which is found to offer best balance of performance improvement and the additional memory overhead.  Each address now in the connection cache map points to a pool of connections. The pool consists of a number connection pool entries. The pool entry is checked out randomly to ensure even usage for the connections for the same address.

2.  Optimization has been done to share the Endpoint for QUIC -- connections from the same address now share the same LazyInitialized endpoint.

3. The quic_client is improved to handle connection errors. Care has been take to avoid the situation where large number of tasks retry creating the connections when running into connection send errors. One connection established will be retried only once across all asynchronous tasks using the connection.

4. In case of tpu-client, it is found that the changes offers 3X of the quic throughput over the result of master code. And for thin-client, the quic throughput is improved about 40% improvement. 

Fixes #

### CODE REVIEW

**Comment by lijunwangs** [comment]
QUIC throughput comparison this branch against master for the same tpu-client run.
./target/release/solana-bench-tps -u http://35.247.120.58:8899 --identity  /home/lijun_solana_com/.config/solana/id.json   --tx_count 1000 --thread-batch-sleep-ms 0   -t 20 --duration 60 --tpu-use-quic -n 35.247.120.58:8000   --read-client-keys /home/lijun_solana_com/gce-keypairs.yaml --use-tpu-client 
 
tpu-client-quic-throughput-compare

**Comment by lijunwangs** [comment]
tpu-client for different pool size, it seems pool-size is sweet spot, more connection pool size beyond does not improvement the throughput significantly.

tpu-client-quic-throughput-different-pool-size

**Comment by lijunwangs** [comment]
Thin-client transaction processing difference between master and this branch as reported by the tool

./target/release/solana-bench-tps -u http://35.247.120.58:8899 --identity  /home/lijun_solana_com/.config/solana/id.json   --tx_count 1000 --thread-batch-sleep-ms 0   -t 20 --duration 60 --tpu-use-quic -n 35.247.120.58:8000   --read-client-keys /home/lijun_solana_com/gce-keypairs.yaml --use-tpu-client 

Master

[2022-06-06T06:05:03.374177459Z INFO  solana_bench_tps::bench] 	Average TPS: 7567.2485
[2022-06-06T06:09:47.120292288Z INFO  solana_bench_tps::bench] 	Average TPS: 1277.8553
[2022-06-06T06:11:37.990492156Z INFO  solana_bench_tps::bench] 	Average TPS: 2094.5317


This Branch

Pool Size: 4

[2022-06-06T06:15:13.781609306Z INFO  solana_bench_tps::bench] 	Average TPS: 15909.017
[2022-06-06T06:17:49.671139249Z INFO  solana_bench_tps::bench] 	Average TPS: 19498.588
[2022-06-06T06:19:28.897689054Z INFO  solana_bench_tps::bench] 	Average TPS: 19016.39

**Comment by lijunwangs** [comment]
QUIC throughput comparison between master, this branch and different connection pool size.  This branch offers significantly better QUIC throughput measured by chunk_received. Again, the pool size =4 seems to be a good balance.
 
thin-client-quic-throughput-compare

File: client/src/connection_cache.rs [hunk 1]
```diff
@@ -215,11 +217,49 @@ impl ConnectionCacheStats {
...
+        let n = rng.gen_range(0, self.connections.len());
+        let connection = self.connections[n].clone();
+        self.references.fetch_add(1, Ordering::Relaxed);
+        connection
+    }
+
+    /// Check if we need to create a new connection. If the count of the connections
+    /// is smaller than the pool size and we have outstanding users of the pool.
+    fn need_new_connection(&self, required_pool_size: usize) -> bool {
+        self.connections.len() < required_pool_size && self.references.load(Ordering::Relaxed) > 0
```
Comment (ryleung-solana) [inline client/src/connection_cache.rs:252]: I think we might need to use acquire ordering here (https://stackoverflow.com/questions/27751025/why-is-an-acquire-barrier-needed-before-deleting-the-data-in-an-atomically-refer). In any case, I'm a little uncomfortable with seemingly reinventing/reimplementing the wheel with reference-counted resources, which is notoriously hard to get right and fast for us mortals 😅...

File: streamer/src/quic.rs [hunk 2]
```diff
@@ -446,12 +451,14 @@ fn handle_connection(
     stake: u64,
 ) {
     tokio::spawn(async move {
-        debug!(
-            "quic new connection {} streams: {} connections: {}",
+        info!(
```
Comment (sakridge) [inline streamer/src/quic.rs:454]: don't think these should be `info` by default

File: streamer/src/quic.rs [hunk 3]
```diff
@@ -537,7 +555,8 @@ pub fn spawn_server(
                 }
 
                 if let Ok(Some(connection)) = timeout_connection {
-                    if let Ok(new_connection) = connection.await {
+                    let result = connection.await;
+                    if let Ok(new_connection) = result {
```
Comment (sakridge) [inline streamer/src/quic.rs:559]: necessary?

File: core/src/banking_stage.rs [hunk 4]
```diff
@@ -558,21 +558,22 @@ impl BankingStage {
...
+                conn.send_wire_transaction_batch_async(packet_vec)
             } else {
                 // All other transactions can be forwarded using QUIC, get_connection() will use
                 // system wide setting to pick the correct connection object.
                 banking_stage_stats
                     .forwarded_transaction_count
                     .fetch_add(packet_vec_len, Ordering::Relaxed);
-                get_connection(&addr)
+                let conn = get_connection(&addr);
+                conn.send_wire_transaction_batch_async(packet_vec)
```
Comment (sakridge) [inline core/src/banking_stage.rs:575]: is there any change here?

File: client/src/connection_cache.rs (hunk 1 above)
Comment (lijunwangs) [reply client/src/connection_cache.rs]: Yes. It is tricky -- that reference count is an optimization. And it turns out the optimization is premature and ineffective in our use case as it is very easy to have multiple threads using the connections in our case driven from the runtime. I have removed all reference counting logic!

File: core/src/banking_stage.rs (hunk 4 above)
Comment (lijunwangs) [reply core/src/banking_stage.rs:575]: Good catch -- I no longer need this change in this file -- after I removed the reference counting changes in the connection pool entry. I will revert it.

**Review by joncinque — COMMENTED** [review]
Changes look good to me (with other comments addressed).  This brings up another idea though: after #25667 lands, maybe the same `ConnectionMap` shouldn't be shared everywhere, and instead each stage should have its own cache so there's no contention around individual connection instances.  After these both land, it's worth a test!

File: streamer/src/quic.rs (hunk 3 above)
Comment (lijunwangs) [reply streamer/src/quic.rs:559]: It was used to dump out the error reason at line 631

File: streamer/src/quic.rs (hunk 2 above)
Comment (lijunwangs) [reply streamer/src/quic.rs:454]: Okay -- I used these info to debug connection issues. I can turn them into debug level.

**Comment by pgarg66** [comment]
> Changes look good to me (with other comments addressed). This brings up another idea though: after #25667 lands, maybe the same `ConnectionMap` shouldn't be shared everywhere, and instead each stage should have its own cache so there's no contention around individual connection instances. After these both land, it's worth a test!

I don't think most stages in a typical node need access to the connection map. It's mainly used by `banking_stage` to forward transactions. QUIC warmup service also uses it (to pre-establish the connections). We have multiple threads in `banking_stage`. So some contention resolution will be needed there.

File: core/src/banking_stage.rs (hunk 4 above)
Comment (lijunwangs) [reply core/src/banking_stage.rs]: done

**Comment by lijunwangs** [comment]
QUIC vs UDP TPS comparison:

lijun_solana_com@lijun-dev:~/solana$ ./target/release/solana-bench-tps -u http://35.247.120.58:8899 --identity  /home/lijun_solana_com/.config/solana/id.json   --tx_count 1000 --thread-batch-sleep-ms 0   -t 20 --duration 60 --tpu-use-quic -n 35.247.120.58:8000   --read-client-keys /home/lijun_solana_com/gce-keypairs.yaml --use-tpu-client --tpu-connection-pool-size 4

QUIC

    Highest TPS: 64556.06 sampling period 1s max transactions: 1035621 clients: 1 drop rate: 0.69
[2022-06-07T16:32:42.677615462Z INFO  solana_bench_tps::bench] 	Average TPS: 17225.367
    Highest TPS: 61725.24 sampling period 1s max transactions: 1116997 clients: 1 drop rate: 0.74
[2022-06-07T16:42:39.902681679Z INFO  solana_bench_tps::bench] 	Average TPS: 18418.15
    Highest TPS: 50606.56 sampling period 1s max transactions: 920453 clients: 1 drop rate: 0.74
[2022-06-07T16:56:38.850249756Z INFO  solana_bench_tps::bench] 	Average TPS: 15272.705



UDP


    Highest TPS: 51976.84 sampling period 1s max transactions: 1134838 clients: 1 drop rate: 0.88
[2022-06-07T17:00:10.819627196Z INFO  solana_bench_tps::bench] 	Average TPS: 18845.572
    Highest TPS: 84689.73 sampling period 1s max transactions: 979811 clients: 1 drop rate: 0.89
[2022-06-07T17:01:40.891839534Z INFO  solana_bench_tps::bench] 	Average TPS: 16227.214
    Highest TPS: 56075.14 sampling period 1s max transactions: 892222 clients: 1 drop rate: 0.90
[2022-06-07T17:05:28.709139026Z INFO  solana_bench_tps::bench] 	Average TPS: 14755.019

**Comment by lijunwangs** [comment]
QUIC vs UDP TPS comparison using thin-client: The TPS seems to be comparable. UDP seems to send more transactions per second than QUIC (notice the higher drop rate) which is probably due to its lower overhead?

lijun_solana_com@lijun-dev:~/solana$ ./target/release/solana-bench-tps -u http://35.247.120.58:8899 --identity /home/lijun_solana_com/.config/solana/id.json --tx_count 1000 --thread-batch-sleep-ms 0 -t 20 --duration 30 --tpu-use-quic -n 35.247.120.58:8000 --rpc-addr 35.247.120.58:8899 --tpu-addr 35.247.120.58:8003 --read-client-keys /home/lijun_solana_com/gce-keypairs.yaml


QUIC

    Highest TPS: 36137.23 sampling period 1s max transactions: 655991 clients: 1 drop rate: 0.67
[2022-06-07T17:18:13.432181280Z INFO  solana_bench_tps::bench] 	Average TPS: 21531.566
    Highest TPS: 38774.46 sampling period 1s max transactions: 653547 clients: 1 drop rate: 0.67
[2022-06-07T17:21:22.756131710Z INFO  solana_bench_tps::bench] 	Average TPS: 21310.664
    Highest TPS: 48156.46 sampling period 1s max transactions: 626314 clients: 1 drop rate: 0.70
[2022-06-07T17:22:34.144676247Z INFO  solana_bench_tps::bench] 	Average TPS: 20735.342


UDP

    Highest TPS: 60185.12 sampling period 1s max transactions: 652289 clients: 1 drop rate: 0.84
[2022-06-07T17:39:57.226081198Z INFO  solana_bench_tps::bench] 	Average TPS: 21640.938
    Highest TPS: 43153.35 sampling period 1s max transactions: 589275 clients: 1 drop rate: 0.87
[2022-06-07T17:40:54.019701934Z INFO  solana_bench_tps::bench] 	Average TPS: 19126.54
    Highest TPS: 44354.58 sampling period 1s max transactions: 579786 clients: 1 drop rate: 0.87
[2022-06-07T17:44:02.417174299Z INFO  solana_bench_tps::bench] 	Average TPS: 19207.791

**Review by sakridge — DISMISSED** [review]
lgtm

**Comment by t-nelson** [comment]
> UDP seems to send more transactions per second than QUIC (notice the higher drop rate) which is probably due to its lower overhead?

any appreciable difference in wire bytes for the QUIC overhead?

**Pull request merged**